### PR TITLE
[DRAMsys 2.0] Terapool DRAMsys Merge Request

### PR DESCRIPTION
**Pull request by yichao-zh**
Summarize the changes concisely

## Changelog
- DRAMsys dynamic library co-simulation support.
- MemPool system level interconnection update for connecting DRAM PHY.

### Added
- DRAMsys2.0 dynamic library.
- System level interconnection and address scrambler.

### CODE REVIEW

**Comment by yichao-zh** [comment]
Hello Guys, this is the first draft for the DRAM work merge. Please check if everything is ok. I will keep update and fix your suggestions in next few days. Thanks!

File: config/config.mk [hunk 1]
```diff
@@ -71,4 +66,4 @@ xpulpimg ?= 1
 
 # This parameter is only used for TeraPool configurations
 num_sub_groups_per_group ?= 1
-remote_group_latency_cycles ?= 7
+remote_group_latency_cycles ?= 7
```
Comment (SamuelRiedel) [inline config/config.mk:69]: Missing newline

File: config/mempool.mk [hunk 2]
```diff
@@ -21,7 +21,17 @@ num_cores_per_tile ?= 4
...
+
+# Number of DMA backends in each group
+dmas_per_group ?= 1 # Brust Length = 16
+
+# L2 Banks/Channels
+l2_size  ?= 4194304  # 400000
+l2_banks ?= 4
+
+# DRAM channel interleaving
+dram_axi_width_interleaved ?= 16
```
Comment (SamuelRiedel) [inline config/mempool.mk:37]: Here we are also missing a newline.

File: config/mempool.mk [hunk 3]
```diff
@@ -21,7 +21,17 @@ num_cores_per_tile ?= 4
 banking_factor ?= 4
 
 # Radix for hierarchical AXI interconnect
-axi_hier_radix ?= 20
+axi_hier_radix ?= 17
```
Comment (SamuelRiedel) [inline config/mempool.mk:24]: I checked, and this change should not impact the backend too much, so if it actually helps with the DRAM's performance, we can keep it that way. But it would be interesting to know why we can't get full performance with smaller bursts on the DRAM.

File: config/minpool.mk [hunk 4]
```diff
@@ -29,11 +29,18 @@ axi_data_width ?= 256
...
-dmas_per_group ?= 1
-
 # Radix for hierarchical AXI interconnect
 axi_hier_radix ?= 2
 
 # Number of AXI masters per group
 axi_masters_per_group ?= 1
+
+# Number of DMA backends in each group
+dmas_per_group ?= 1 # Brust Length = 16
```
Comment (SamuelRiedel) [inline config/minpool.mk:39]: In MinPool, the burst length will be 8, right?

File: hardware/dram_rtl_sim/Bender.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2021 ETH Zurich and University of Bologna.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+package:
```
Comment (SamuelRiedel) [inline hardware/dram_rtl_sim/Bender.yml:5]: It looks like `dram_rtl_sim` is a full dependency. I think it would make more sense to create a dedicated repository for it that can be included as a dependency instead of flattening it in MemPool. Especially since it's almost 200 files and more than 2 million lines of code. :astonished: Is it also used in other projects or just here?

File: hardware/dram_rtl_sim/dramsys_lib/resources/traces/chstone-adpcm_32.stl [hunk 6]
```diff
@@ -0,0 +1,1211 @@
+31:	read	0x400140
```
Comment (SamuelRiedel) [inline hardware/dram_rtl_sim/dramsys_lib/resources/traces/chstone-adpcm_32.stl:1]: What are those traces? Do we need them in here?

File: hardware/dram_rtl_sim/dramsys_lib/libDRAMSysLibrary.so
Comment (SamuelRiedel) [inline hardware/dram_rtl_sim/dramsys_lib/libDRAMSysLibrary.so:1]: We should not check in precompiled libraries. Is there a way to compile this library easily instead?

File: hardware/scripts/dramsys_plot.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
```
Comment (SamuelRiedel) [inline hardware/scripts/dramsys_plot.py:1]: Is this script MemPool specific, or could it also go into the `dram_rtl_sim` dependency and be used for other projects as well?

File: hardware/src/mempool_cc.sv [hunk 8]
```diff
@@ -206,8 +206,8 @@ module mempool_cc
   logic [63:0] cycle;
   int unsigned stall, stall_ins, stall_raw, stall_lsu, stall_acc;
 
-  always_ff @(posedge rst_i) begin
-    if(rst_i) begin
+  always_ff @(negedge rst_i) begin
+    if(!rst_i) begin
```
Comment (SamuelRiedel) [inline hardware/src/mempool_cc.sv:210]: Why did you change this?

File: hardware/src/mempool_group.sv [hunk 9]
```diff
@@ -1025,8 +1025,8 @@ module mempool_group
         .DataWidth       (AxiDataWidth   ),
         .AddrWidth       (AddrWidth      ),
         .IdWidth         (AxiTileIdWidth ),
-        .AxReqFifoDepth  (2              ),
-        .TransFifoDepth  (1              ),
+        .AxReqFifoDepth  (8              ),
+        .TransFifoDepth  (4              ),
```
Comment (SamuelRiedel) [inline hardware/src/mempool_group.sv:1029]: Are those changes really necessary to get the best performance? Those FIFOs are very expensive in hardware.

File: hardware/src/mempool_sub_group.sv [hunk 10]
```diff
@@ -514,8 +514,8 @@ module mempool_sub_group
       .DataWidth       (AxiDataWidth   ),
       .AddrWidth       (AddrWidth      ),
       .IdWidth         (AxiTileIdWidth ),
-      .AxReqFifoDepth  (2              ),
-      .TransFifoDepth  (1              ),
+      .AxReqFifoDepth  (8              ),
+      .TransFifoDepth  (4              ),
```
Comment (SamuelRiedel) [inline hardware/src/mempool_sub_group.sv:518]: Same question here.

File: hardware/tb/mempool_tb.sv [hunk 11]
```diff
@@ -30,7 +30,7 @@ module mempool_tb;
   localparam BootAddr = 0;
   `endif
 
-  localparam ClockPeriod = 2ns;
+  localparam ClockPeriod = 1ns;
```
Comment (SamuelRiedel) [inline hardware/tb/mempool_tb.sv:33]: Why this change?

File: hardware/tb/mempool_tb.sv [hunk 12]
```diff
@@ -345,6 +350,46 @@ module mempool_tb;
...
+                dut.gen_drams[bank].i_axi_dram_sim.i_sim_dram.load_to_dram(dram_ctrl_info.dram_ctrl_addr, buffer[i]);
+              end
+            end
+          end else begin
+            $display("Cannot initialize address %x, which doesn't fall into the L2 DRAM region.", address);
+          end
+        end
+      end
+    end : l2_init
+  end : gen_drams_init
```
Comment (SamuelRiedel) [inline hardware/tb/mempool_tb.sv:389]: It looks like most of this block is the same as for the L2. Could we merge them and have the ifdef only in the actual preloading in the innermost loops?

File: hardware/Makefile [hunk 13]
```diff
@@ -48,6 +48,13 @@ python          ?= python3
 # Enable tracing
 snitch_trace    ?= 0
 
+# Path to DRAMsyslib
+dramsys_resouces_path ?= $(MEMPOOL_DIR)/hardware/dram_rtl_sim/dramsys_lib/resources
+dramsys_simconfig_path ?= $(dramsys_resouces_path)/simulations/HBM2E-3600.json
+dramsys_lib_path ?= $(MEMPOOL_DIR)/hardware/dram_rtl_sim/dramsys_lib
+questa_args += +DRAMSYS_RES=$(dramsys_resouces_path) +DRAMSYS_CONF=$(dramsys_simconfig_path)
+questa_args += -sv_lib $(dramsys_lib_path)/libDRAMSysLibrary
+
```
Comment (SamuelRiedel) [inline hardware/Makefile:57]: Does this also work if dram is disabled, or should we only add those questa_args if `dram` is defined?

**Review by SamuelRiedel — COMMENTED** [review]
Thanks a lot, Yichao and Chi. In general, it looks good to me, but I think we should discuss putting all the dram simulation stuff in its own repository and including it in systems that need it as a submodule. There are also some linting errors and missing newlines.
